src/ast/passes/codegen_llvm.cpp

```diff
@@ -2372,12 +2376,60 @@ void CodegenLLVM::generateProbe(Probe &probe,
         func_type, section_name, current_attach_point_->address, index);
 }
 
+void CodegenLLVM::visit(Subprog &subprog)
+{
+  std::vector<llvm::Type *> arg_types;
+  // First argument is for passing ctx pointer for output, rest are proper
+  // arguments to the function
+  arg_types.push_back(b_.getInt8PtrTy());
+  std::transform(subprog.args->begin(),
+                 subprog.args->end(),
+                 std::back_inserter(arg_types),
+                 [this](SubprogArg *arg) { return b_.GetType(arg->type); });
+  FunctionType *func_type = FunctionType::get(b_.GetType(subprog.return_type),
+                                              arg_types,
+                                              0);
+
+  Function *func = Function::Create(
+      func_type, Function::InternalLinkage, subprog.name(), module_.get());
+  BasicBlock *entry = BasicBlock::Create(module_->getContext(), "entry", func);
+  b_.SetInsertPoint(entry);
+
+  variables_.clear();
+  ctx_ = func->arg_begin();
+  inside_subprog_ = true;
+
+  int arg_index = 0;
+  for (SubprogArg *arg : *subprog.args) {
+    auto alloca = b_.CreateAllocaBPF(b_.GetType(arg->type), arg->name());
+    b_.CreateStore(func->getArg(arg_index + 1), alloca);
+    variables_.insert({ arg->name(), alloca });
+    ++arg_index;
+  }
+
+  for (Statement *stmt : *subprog.stmts) {
+    auto scoped_del = accept(stmt);
+  }
+  if (subprog.return_type.IsVoidTy())
+    createRet();
+
+  FunctionPassManager fpm;
+  FunctionAnalysisManager fam;
+  llvm::PassBuilder pb;
+  pb.registerFunctionAnalyses(fam);
+  fpm.addPass(UnreachableBlockElimPass());
+  fpm.run(*func, fam);
+}
+
 void CodegenLLVM::createRet(Value *value)
 {
   // If value is explicitly provided, use it
   if (value) {
     b_.CreateRet(value);
     return;
+  } else if (inside_subprog_) {
+    b_.CreateRetVoid();
```

Comment:
I'm assuming this is just a temporary thing until your next PR?

Reply:
This is because unlike with probes, there is not any default return value for subprogs. createRet() without argument is only used to end unreachable code that is then removed in subprogs.